### PR TITLE
Adds beta command `runme env store snapshot`

### PR DESCRIPTION
**Pull request by pastuxso**
This PR supports interacting with the **MonitorEnvStore** in the CLI.

Usage:

`runme env store snapshot --tlsdir=<RUNME TLS CERTIFICATES> --address=<RUNME SERVER ADDRESS> --session=<RUNME SESSION ID>`

The following variables will be used by default if you don't pass arguments:
 
`RUNME_SESSION`
`RUNME_SERVER_ADDR`
`RUNME_TLS_DIR`

Example:

`runme env store snapshot`

image

Issue: https://github.com/stateful/runme/issues/528

### CODE REVIEW

File: internal/runner/session.go [hunk 1]
```diff
@@ -41,6 +41,9 @@ func NewSession(envs []string, proj *project.Project, logger *zap.Logger) (*Sess
 func NewSessionWithStore(envs []string, proj *project.Project, owlStore bool, logger *zap.Logger) (*Session, error) {
 	sessionEnvs := []string(envs)
 
+	id := ulid.GenerateID()
+	sessionEnvs = append(sessionEnvs, fmt.Sprintf("RUNME_SESSION=%s", id))
```
Comment (pastuxso) [inline internal/runner/session.go:45]: I added the RUNME_SESSION variable to infer its value later.

image

File: internal/cmd/environment.go [hunk 2]
```diff
@@ -20,12 +29,103 @@ func environmentCmd() *cobra.Command {
...
+				}
+
+				return errors.Wrap(table.Render(), "failed to render")
+			}
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVar(&addr, "address", "", "The Server address to connect to, i.e. 127.0.0.1:7865")
```
Comment (pastuxso) [inline internal/cmd/environment.go:122]: I'm wondering how to infer the connection when you run the command inside a notebook.

File: internal/runner/service.go [hunk 3]
```diff
@@ -86,6 +91,8 @@ func (r *runnerService) CreateSession(ctx context.Context, req *runnerv1.CreateS
 
 	envs := make([]string, len(req.Envs))
 	copy(envs, req.Envs)
+	envs = append(envs, fmt.Sprintf("RUNME_RUNNER_ADDRESS=%s", r.address))
+	envs = append(envs, fmt.Sprintf("RUNME_RUNNER_TLS_DIR=%s", r.tlsDir))
```
Comment (pastuxso) [inline internal/runner/service.go:95]: It allows us to infer the server params to perform the runme env store snapshot

image

File: internal/cmd/environment.go [hunk 4]
```diff
@@ -20,12 +29,103 @@ func environmentCmd() *cobra.Command {
...
+
+func storeSnapshotCmd() *cobra.Command {
+	var (
+		addr      string
+		tlsDir    string
+		sessionID string
+	)
+
+	cmd := cobra.Command{
+		Use:   "snapshot",
```
Comment (sourishkrout) [inline internal/cmd/environment.go:59]: @pastuxso there's a way to set a `Hidden: true` or something like it. Let's do that please as long as it experimental.

File: internal/runner/service.go [hunk 5]
```diff
@@ -803,8 +816,8 @@ func convertToMonitorEnvStoreResponse(msg *runnerv1.MonitorEnvStoreResponse, sna
 			OriginalValue: item.Value.Original,
 			ResolvedValue: item.Value.Resolved,
 			Status:        status,
-			CreateTime:    item.Var.Created.String(),
-			UpdateTime:    item.Var.Updated.String(),
+			CreateTime:    item.Var.Created.Format(time.RFC3339),
```
Comment (pastuxso) [inline internal/runner/service.go:819]: @sourishkrout let's use the RFC3339 which is more standard for sharing dates through systems and compatible with ISO8601, It was tested with the extension and is working fine.

File: internal/env/env.go [hunk 6]
```diff
@@ -13,3 +18,22 @@ func ConvertMapEnv(mapEnv map[string]string) []string {
 
 	return result
 }
+
+func CleaSessionVars(envs []string) []string {
```
Comment (sourishkrout) [inline internal/env/env.go:22]: Clean?

File: internal/cmd/environment.go [hunk 7]
```diff
@@ -20,12 +30,82 @@ func environmentCmd() *cobra.Command {
...
+	var (
+		addr      string
+		tlsDir    string
+		sessionID string
+	)
+
+	cmd := cobra.Command{
+		Hidden: true,
+		Use:    "snapshot",
+		Short:  "Dump environment variables to stdout",
```
Comment (sourishkrout) [inline internal/cmd/environment.go:63]: Copy & paste description leftover?

File: internal/cmd/server.go [hunk 8]
```diff
@@ -104,7 +104,7 @@ The kernel is used to run long running processes like shells and interacting wit
 			parserv1.RegisterParserServiceServer(server, editorservice.NewParserServiceServer(logger))
 			projectv1.RegisterProjectServiceServer(server, projectservice.NewProjectServiceServer(logger))
 			if enableRunner {
-				runnerServicev1, err := runner.NewRunnerService(logger)
+				runnerServicev1, err := runner.NewRunnerService(logger, addr, tlsDir)
```
Comment (sourishkrout) [inline internal/cmd/server.go:107]: don't quite understand why we need this

File: internal/cmd/environment.go [hunk 9]
```diff
@@ -49,6 +129,46 @@ func environmentDumpCmd() *cobra.Command {
...
+		}
+
+		table.AddField(env.Name)
+		table.AddField(value)
+		table.AddField(env.Spec)
+		table.AddField(env.Origin)
+
+		t, err := time.Parse(time.RFC3339, env.UpdateTime)
+		if err == nil {
+			table.AddField(t.Format(time.RFC1123Z))
```
Comment (sourishkrout) [inline internal/cmd/environment.go:161]: maybe just `t.Format(time.DateTime)` here? on the client it's okay to show "local time"

**Review by sourishkrout — CHANGES_REQUESTED** [review]
Can we "remove" the session injection for now? Don't need it for the demo and it needs more thought.

File: internal/cmd/server.go (hunk 8 above)
Comment (pastuxso) [reply internal/cmd/server.go:107]: They were added to be available later as environment variables when a new session is created.

Comment (pastuxso) [reply internal/cmd/server.go:107]: https://github.com/stateful/runme/pull/532/files#diff-50b92da422aa4c0c738d7a625064364d14c03cafd8b1048eefd6fb9ae10227d9R96

Comment (pastuxso) [reply internal/cmd/server.go:107]: NVM, Now I understand that variables are being injected in the extension.

image

Comment (pastuxso) [reply internal/cmd/server.go:107]: Removing, you're right.

**Comment by pastuxso** [comment]
> Can we "remove" the session injection for now? Don't need it for the demo and it needs more thought.

Yes, removed.

**Review by sourishkrout — APPROVED** [review]
👍 Works. Added some flags.

Likely needs a refactor to use the Runner client interface for non-hidden release which we can do later.

**Pull request merged**